src/lib/shell/commands/Ota.cpp

```diff
@@ -0,0 +1,420 @@
+    }
+
+    if (exchangeCtx == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "Failed to allocate exchange");
+        return;
+    }
+
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlags = TransferControlFlags::kReceiverDrive;
+    initOptions.MaxBlockSize     = kBlockSize;
+    initOptions.FileDesLength    = resourceName.size();
+    initOptions.FileDesignator   = reinterpret_cast<const uint8_t *>(resourceName.data());
+
+    sDfuManager.SetInitialExchange(exchangeCtx);
+    CHIP_ERROR error = sDfuManager.InitiateTransfer(&DeviceLayer::SystemLayer(), TransferRole::kReceiver, initOptions,
+                                                    System::Clock::Seconds16(20));
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "Failed to initiate BDX transfer: %" CHIP_ERROR_FORMAT, error.Format());
+    }
+}
+
+void OnQueryImageFailure(void * /* context */, EmberAfStatus status)
+{
+    ChipLogError(SoftwareUpdate, "QueryImage failed: %" PRIu16, static_cast<uint16_t>(status));
+}
+
+void OnQueryImageConnection(void * /* context */, DeviceProxy * deviceProxy)
+{
+    // Initialize cluster object
+    Controller::OtaSoftwareUpdateProviderCluster cluster;
+    CHIP_ERROR error = cluster.Associate(deviceProxy, sOtaContext.providerEndpointId);
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "Associate failed: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    // Send QueryImage command
+    uint16_t vendorId        = 0;
+    uint16_t productId       = 0;
+    uint16_t hardwareVersion = 0;
+    uint16_t softwareVersion = 0;
+
+    DeviceLayer::ConfigurationMgr().GetVendorId(vendorId);
+    DeviceLayer::ConfigurationMgr().GetProductId(productId);
+    DeviceLayer::ConfigurationMgr().GetProductRevision(hardwareVersion);
+    DeviceLayer::ConfigurationMgr().GetFirmwareRevision(softwareVersion);
+
+    QueryImage::Type request;
+    request.vendorId           = static_cast<VendorId>(vendorId);
+    request.productId          = productId;
+    request.softwareVersion    = softwareVersion;
+    request.protocolsSupported = kRequestorProtocols;
+    request.hardwareVersion.SetValue(hardwareVersion);
+    request.location.Emplace(kRequestorLocation);
+    request.requestorCanConsent.SetValue(kRequestorCanConsent);
+    error = cluster.InvokeCommand(request, /* context */ nullptr, OnQueryImageResponse, OnQueryImageFailure);
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "QueryImage failed: %" CHIP_ERROR_FORMAT, error.Format());
+    }
+}
+
+void OnApplyUpdateResponse(void * context, const ApplyUpdateResponse::DecodableType & response)
+{
+    ChipLogProgress(SoftwareUpdate, "Received ApplyUpdate response: %" PRIu16, static_cast<uint16_t>(response.action));
+
+    switch (response.action)
+    {
+    case EMBER_ZCL_OTA_APPLY_UPDATE_ACTION_PROCEED: {
+        CHIP_ERROR error = sDfuManager.ApplyUpdate();
```

Comment:
Is this a blocking call? Will this trigger a reboot immediately? If either is true, this should be changed to schedule the work for later and let this handler function return.

Same goes for the `DiscardUpdate()` call below.

Reply:
It's not blocking. It just marks the image as "ready to be swapped on the next reboot". And the reboot is not scheduled automatically for now, but I consider the DFUManager class temporary anyway.